File: src/lib/ai.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { Config } from "./types";

/** OpenAI 多模态 content part */
export interface ImagePart {
  type: "image_url";
  image_url: { url: string }; // data: URL
}
export interface TextPart {
  type: "text";
  text: string;
}

export interface AiMessage {
  role: "system" | "user" | "assistant";
  content: string | (TextPart | ImagePart)[];
}

/**
 * 流式对话：invoke 立即返回，增量经 `ai-chunk-{id}` 事件到达。
 * onDelta 携带每个增量片段；返回完整文本。
 */
export async function aiChat(
  cfg: Config,
  messages: AiMessage[],
  onDelta: (delta: string, full: string) => void,
  maxTokens?: number,
  onReasoningDelta?: (reasoningFull: string) => void
): Promise<string> {
  const requestId = `${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
  return new Promise((resolve, reject) => {
    let full = "";
    let reasoningFull = "";
    let un: UnlistenFn | null = null;
    let settled = false;
    const finish = (ok: boolean, err?: string) => {
      if (settled) return;
      settled = true;
      un?.();
      if (ok) resolve(full);
      else reject(new Error(err || "AI 请求失败"));
    };
    listen<{ delta: string; reasoning: boolean; done: boolean; error?: string }>(
      `ai-chunk-${requestId}`,
      (e) => {
        const { delta, reasoning, done, error } = e.payload;
        if (error) return finish(false, error);
        if (delta && !reasoning) {
          // 推理模型的思维链（reasoning）不进正文
          full += delta;
          onDelta(delta, full);
        }
        if (delta && reasoning) {
          reasoningFull += delta;
          onReasoningDelta?.(reasoningFull);
        }
        if (done) finish(true);
      }
    ).then((u) => {
      un = u;
      invoke("ai_chat", {
        requestId,
        baseUrl: cfg.aiBaseUrl,
        apiKey: cfg.aiApiKey,
        model: cfg.aiModel,
        messages,
        maxTokens: maxTokens ?? null,
      }).catch((err) => finish(false, String(err)));
    });
  });
}

File: src-tauri/src/ai.rs
use futures_util::StreamExt;
use serde::{Deserialize, Serialize};
use tauri::Emitter;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ChatMsg {
    pub role: String,
    /// 纯文本 string 或 OpenAI 多模态数组 [{type:"text"},{type:"image_url",...}]
    pub content: serde_json::Value,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct AiChunk {
    delta: String,
    /// 推理模型的思维链增量（deepseek-v4-flash-vision-exp 等）
    reasoning: bool,
    done: bool,
    error: Option<String>,
}

/// AI 对话（OpenAI 兼容 /chat/completions，SSE 流式）。
/// 通过事件 `ai-chunk-{requestId}` 推送增量；前端拼接渲染。
/// 流式请求在 tokio 任务中执行，command 立即返回，UI 不阻塞。
#[tauri::command]
pub async fn ai_chat(
    app: tauri::AppHandle,
    request_id: String,
    base_url: String,
    api_key: String,
    model: String,
    messages: Vec<ChatMsg>,
    max_tokens: Option<u32>,
) -> Result<(), String> {
    if api_key.is_empty() {
        return Err("未配置 API Key，请先在设置中填写".to_string());
    }
    let url = format!("{}/chat/completions", base_url.trim_end_matches('/'));

    tokio::spawn(async move {
        let event = format!("ai-chunk-{request_id}");
        let emit = |delta: String, reasoning: bool, done: bool, error: Option<String>| {
            let _ = app.emit(&event, AiChunk { delta, reasoning, done, error });
        };

        let client = match reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(300))
            .build()
        {
            Ok(c) => c,
            Err(e) => {
                emit(String::new(), false, true, Some(e.to_string()));
                return;
            }
        };

        let mut body = serde_json::json!({ "model": model, "messages": messages, "stream": true });
        if let Some(mt) = max_tokens {
            body["max_tokens"] = serde_json::json!(mt);
        }

        let resp = match client
            .post(&url)
            .header("Authorization", format!("Bearer {api_key}"))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
        {
            Ok(r) => r,
            Err(e) => {
                emit(String::new(), false, true, Some(format!("网络错误：{e}")));
                return;
            }
        };

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            emit(String::new(), false, true, Some(format!("HTTP {status}：{}", truncate(&text, 500))));
            return;
        }

        let mut stream = resp.bytes_stream();
        let mut buf = String::new();
        loop {
            match stream.next().await {
                Some(Ok(chunk)) => {
                    buf.push_str(&String::from_utf8_lossy(&chunk));
                    // SSE：按行解析 data:
                    while let Some(pos) = buf.find('\n') {
                        let line: String = buf.drain(..=pos).collect();
                        let line = line.trim();
                        if !line.starts_with("data:") {
                            continue;
                        }
                        let data = line[5..].trim();
                        if data == "[DONE]" {
                            emit(String::new(), false, true, None);
                            return;
                        }
                        if let Ok(v) = serde_json::from_str::<serde_json::Value>(data) {
                            // 推理模型：思维链在 reasoning_content，正文在 content，都要透传
                            let reasoning = v
                                .pointer("/choices/0/delta/reasoning_content")
                                .and_then(|c| c.as_str())
                                .unwrap_or("");
                            if !reasoning.is_empty() {
                                emit(reasoning.to_string(), true, false, None);
                            }
                            let delta = v
                                .pointer("/choices/0/delta/content")
                                .and_then(|c| c.as_str())
                                .unwrap_or("");
                            if !delta.is_empty() {
                                emit(delta.to_string(), false, false, None);
                            }
                        }
                    }
                }
                Some(Err(e)) => {
                    emit(String::new(), false, true, Some(format!("流中断：{e}")));
                    return;
                }
                None => {
                    emit(String::new(), false, true, None);
                    return;
                }
            }
        }
    });

    Ok(())
}

fn truncate(s: &str, n: usize) -> String {
    s.chars().take(n).collect()
}
